### DISCUSSION TITLE
Using with as in collection view

### DISCUSSION THREAD

**Discussion by starredev** [pocketbase/pocketbase#5247] [Q&A]
is WITH AS supported? I'm trying alot but it seems it is not implemented/working 

forexample:

WITH places_data AS (
    SELECT 
        json_extract(places.value, '$.safety_measures') AS safety_measures
    FROM 
        workbooks,
        json_each(json_column) AS places
)

**Comment by ganigeorgiev** [comment]
It should be supported but the above query doesn't look correct for me (or at least I don't understand what you are trying to do). At first glance it seems that it is missing a SELECT clause that will use the WITH data and at minimum it doesn't provide a list of valid select columns that should be converted to collection fields.

The easiest way to rule out PocketBase issue is to try to run the query directly against the `pb_data/data.db` file:

```sh
sqlite3 pb_data/data.db 'yourQuery'
```

Also keep in mind the View collections caveats noted in the Admin UI:

> - Wildcard columns (*) are not supported.
> - The query must have a unique id column.
>  If your query doesn't have a suitable one, you can use the universal (ROW_NUMBER() OVER()) as id.
> - Expressions must be aliased with a valid formatted field name (eg. MAX(balance) as maxBalance).

If you need further help, please export your collections configuration as json (_Admin UI > Settings > Export collections_) and elaborate what the above query is supposed to do.

**Reply by starredev** [reply]
> It should be supported but the above query doesn't look correct for me (or at least I don't understand what you are trying to do). At first glance it seems that it is missing a SELECT clause that will use the WITH data and at minimum it doesn't provide a list of valid select columns that should be converted to collection fields.
> 
> The easiest way to rule out PocketBase issue is to try to run the query directly against the `pb_data/data.db` file:
> 
> ```shell
> sqlite3 pb_data/data.db 'yourQuery'
> ```
> 
> Also keep in mind the View collections caveats noted in the Admin UI:
> 
> > * Wildcard columns (*) are not supported.
> > * The query must have a unique id column.
> >   If your query doesn't have a suitable one, you can use the universal (ROW_NUMBER() OVER()) as id.
> > * Expressions must be aliased with a valid formatted field name (eg. MAX(balance) as maxBalance).
> 
> If you need further help, please export your collections configuration as json (_Admin UI > Settings > Export collections_) and elaborate what the above query is supposed to do.

Thanks @ganigeorgiev sqlite3 pb_data did the trick to see exactly what I am doing wrong. Had an , (comma) after COUNT AS which was not correct. The query is quite complex but it works now :)

```
WITH extracted_places AS (
    SELECT
        id as workbookId,
        json_extract(plan, '$.Instructie BD.location_safety_instruction.places') AS places
    FROM
        workbooks
),
places AS (
    SELECT
        ep.workbookId,
        json_each.value AS place
    FROM
        extracted_places ep,
        json_each(ep.places) AS json_each
),
safety_measures AS (
    SELECT
        p.workbookId,
        json_extract(p.place, '$.safety_measures') AS safety_measures
    FROM
        places p
),
measures AS (
    SELECT
        sm.workbookId,
        json_each.key AS key_name,
        json_each.value AS value
    FROM
        safety_measures sm,
        json_each(sm.safety_measures) AS json_each
),
keys_measures AS (
    SELECT
        m.workbookId,
        json_each.value AS value,
        CASE
            WHEN json_each.key = 'a' THEN 'key a'
            WHEN json_each.key = 'b' THEN 'key b'
            WHEN json_each.key = 'c' THEN 'key c'
        END AS key_name
    FROM
        measures m,
        json_each(m.value) AS json_each
    WHERE m.key_name = 'keys'
),
counts AS (
    SELECT
        m.workbookId,
        m.key_name,
        m.value,
        CASE
            WHEN json_valid(m.value) AND json_type(m.value) = 'array' THEN json_array_length(m.value)
            WHEN json_valid(m.value) AND json_type(m.value) = 'object' THEN (
                SELECT COUNT(*)
                FROM json_each(m.value) AS json_each
            )
            ELSE 1 -- Assume scalar values count as 1
        END AS item_count
    FROM
        measures m
    UNION ALL
    SELECT
        k.workbookId,
        k.key_name,
        k.value,
        json_array_length(k.value) AS item_count
    FROM
        keys_measures k
)
SELECT
    (ROW_NUMBER() OVER()) AS id,
    workbookId,
    key_name,
    SUM(item_count) AS count
FROM counts
WHERE key_name != 'keys'
GROUP BY workbookId, key_name
ORDER BY workbookId;
```
